programs/survival/horse_buddy.sc

```diff
@@ -0,0 +1,306 @@
+global_training = {
+  'uuid' -> null,
+  'food' -> 'wheat',
+  'generic.movement_speed' -> 0,
+  'horse.jump_strength' -> 0,
+  'generic.max_health' -> 0
+};
+
+// Rest the training when ridding a new horse
+// or after rolling for new stats.
+_rest_training(uuid) -> (
+  global_training = {
+    'uuid' -> uuid,
+    'food' -> _random_food(),
+    'generic.movement_speed' -> 0,
+    'horse.jump_strength' -> 0,
+    'generic.max_health' -> 0
+  };
+);
+
+
+// Roll an amount to add or remove from a given stat
+_roll_stat_for(stat, modifier) -> (
+  call(
+    // Adjust movement speed by 10% of vanilla range
+    { 'generic.movement_speed' -> _(x) -> (
+      0.225 * 0.2 * ((x) - rand(1))
+    ),
+    // Adjust jump height by roughly 1/2 blocks
+    'horse.jump_strength' -> _(x) -> (
+      0.25 * ((x) - rand(1))
+    ),
+    // Adjust health by 1/2 Hearts
+    'generic.max_health' -> _(x) -> (
+      3.0 - floor(rand(6)+((x-0.5)*6))
+    )
+  }:stat, modifier)
+);
+
+_clamp(mn, mx, val) -> max(mn, min(mx, val));
+
+
+
+//--- FOOD ---//
+
+_random_food() -> keys(_food_strengths()):(floor(rand(6)));
+
+_food_strengths() -> ({
+  'wheat'->1,
+  'apple'->2,
+  'golden_carrot'->6,
+  'golden_apple'->64,
+  'hay_block'->12
+});
+
+
+//--- ON PLAYER METHODS ---//
+
+__on_player_rides(p, forward, strafe, jumping, sneaking) -> (
+  global_training:'horse.jump_strength' += number(jumping);
+  global_training:'generic.movement_speed' += ceil(forward) + ceil(strafe);
+);
+
+__on_player_interacts_with_entity(p, horse, hand) -> (
+  kind = horse~'type';
+  if(kind == 'horse' || kind == 'donkey',
+    _horse_logic(p, horse);
+  );
+);
+
+
+//--- LOGIC n stuff ---//
+
+_horse_logic(p, horse) -> (
+
+    // reset training for new horses
+    if(horse~'uuid' != global_training:'uuid',
+      _rest_training(horse~'uuid');
+    );
+
+
+    item = p~'holds';
+    // when you feed the demanding horse what it wants
+    if(item:0 == global_training:'food',
+
+      // Pick a new craving
+      global_training:'generic.max_health' += _food_strengths():(item:0);
+      global_training:'food' = _random_food();
+
+      particle('happy_villager', horse~'pos'+[0,1.5,0], 8, 0.5);
+
+      // once the horse is satisfied, re-roll some stats.
+      if(global_training:'generic.max_health' > 16,
+        _roll_new_stats(horse);
+        schedule(10, '_after_hint', p)
+      );
+
+    //elif make the horse complain about its food.
+    ,_food_strengths():(item:0) || !rand(8),
+      schedule(10, '_food_hint', p)
+
+    //elif give random hits about potions and things.
+    ,!rand(5), 
+      schedule(10, '_random_hint', p)
+
+    );
+);
+
+//--- ROLL NEW STATS ---//
+
+_roll_new_stats(horse) -> (
+  effect_mod = _effect_stat_mods(reduce(horse~'effect', _a += _:0, []));
+  attributes = horse~'attribute';
+
+  // these are the magic numbers that try to balance out the training
+  // we need to turn numbers like 500 and 6000 into reasonable stat modifiers no larger than 1.
```

Comment:
Can u figure out the largest possible value these numbers could have? If so you could take the log and then scale them (so smaller numbers also have some significance at that scale)
If not, just do log (base 10) and divide by like 4 (cos that way 10,000 becomes 1 and 10 becomes 0.25)

Reply:
Ok I tired the log stuff. The problem is I actually just want the effort to be linear.
Any exponential effects I would want to be slight.
And getting exponential math to feel right in game is more effort.